### DISCUSSION TITLE
Android Library for Printing Features?

### DISCUSSION THREAD

**Discussion by zeitwesen** [NaitLee/Cat-Printer#79] [Ideas]
Hi,

Thanks for your project - it works great. But as I have built my own Android-App to cover and automate my daily hustles in digital life I though of a feature to "just print" some photos via "Sharing" in Android. For that I am searching for a library, that I can easily integrate into my own Android app project. Do you plan to port/wrap your "printer communication" code so it has a nice API so I can use it inside my app? 

Maybe other use-cases arise from that - e.g. use it as a receipt printer in an account app, printing notes, etc... 

What do you all think?

**Comment by NaitLee** [comment]
Though this codebase may be used like a library, it isn’t specifically designed for Android, as there are just some Python code.

That said, if you have ability to develop for Android, it should be easy to program the [printers’ protocol](https://github.com/NaitLee/Cat-Printer/blob/9a29e401fcf9c004c2b949ffd5fae4be679fcd95/printer_lib/commander.py#L60), produce data with it, and send to devices via Android API for bluetooth. ([also some code to interact with printer](https://github.com/NaitLee/Cat-Printer/blob/9a29e401fcf9c004c2b949ffd5fae4be679fcd95/printer.py#L330))

With some [references](https://github.com/NaitLee/Cat-Printer#credits) and a library (for bluetooth), that just costed me 3 days or so in Python. I’m novice to such “project” that time.

I’d also share something for connection to the printer. Any other problem just ask :)

---

These printers are Bluetooth Low Energy (BLE) peripherals.
They behave as servers. Once a client (your PC or mobile phone) connects to it, some *services* are provided for interaction.
Each service have some *characteristics*, that you can *read and/or write* data to, or receive *notification* data from.
Both *services* and *characteristics* are referenced by their UUIDs.

When scanning for BLE devices, the devices may include their service they provide in the announcement data.
This can be used to filter out devices, to take only what we want.

For these printers, just care about these:

| Description | UUID |
| --- | --- |
| Announced service UUID to *filter out* supported printers | `0000af30-0000-1000-8000-00805f9b34fb` |
| Service UUID to *get* after connection to device’s GATT server | `0000ae30-0000-1000-8000-00805f9b34fb` |
| Characteristic to *write* produced data to (`writeWithoutResponse`) | `0000ae01-0000-1000-8000-00805f9b34fb` |
| Characteristic to receive pause/resume *notification* from | `0000ae02-0000-1000-8000-00805f9b34fb` |

(Note: yes, the printer’s announced service doesn’t correspond to what it really have, we can just ignore that fact)
(Also, don’t filter printers by their device name. It’s a bad design that I have to admit.)

I’m planning to make another frontend using Web Bluetooth API, using all of these. That may serve a better example in the future, in case this codebase couldn’t convince.

**Comment by NaitLee** [comment]
The new project [kitty-printer](https://github.com/NaitLee/kitty-printer) serves an example, especially [this file](https://github.com/NaitLee/kitty-printer/blob/main/common/cat-protocol.ts).
